### PR TITLE
Add Crystal.main

### LINKED ISSUE #4793: Feature: Allow redefining main with a different name exposed in Crystal
I'm writing a binding to [Allegro](http://liballeg.org/), and running into a crash.  I've created a slim down reproducible script:

```crystal
@[Link("allegro")]
lib LibAllegro
  fun al_run_main(argc : LibC::Int, argv : LibC::Char**, user_main : (LibC::Int, LibC::Char** -> LibC::Int)) : LibC::Int
end

module CrystalAllegro
  @@main : Proc(Nil)?

  def self.run_main(&@@main)
    LibAllegro.al_run_main(0, nil, ->run_main_callback)
  end

  private def self.run_main_callback(_argc : LibC::Int, _argv : LibC::Char**) : LibC::Int
    @@main.not_nil!.call
    0
  rescue ex
    STDERR.puts ex.inspect_with_backtrace
    1
  end
end

module Example
  def self.main
    5000.times { |i| puts i }
  end
end

CrystalAllegro.run_main { Example.main }
```

The result is one of the following

```
...
892
893
[0x10d4624eb] *CallStack::print_backtrace:Int32 +107
[0x10d44de8c] __crystal_sigfault_handler +60
[0x7fffa0d1eb3a] _sigtramp +26
[0x10d46633c] *Fiber::switch_stacks<Pointer(Pointer(Void)), Pointer(Pointer(Void))>:Nil +12
```
```
# Same as above but with the extra line about signal 11
...
1132
1133
Invalid memory access (signal 11) at address 0x28
[0x10978b4eb] *CallStack::print_backtrace:Int32 +107
[0x109776e8c] __crystal_sigfault_handler +60
[0x7fffa0d1eb3a] _sigtramp +26
[0x10978f33c] *Fiber::switch_stacks<Pointer(Pointer(Void)), Pointer(Pointer(Void))>:Nil +12
```
```
...
601
602
# and then it just stops...nothing else
```
```
# If I use a slightly more complicated main function (although it doesn't seem to happen with this particular example)
Collecting from unknown thread
Program received and didn't handle signal IOT (6)
```

Some notes:
- Crystal 0.23.1 (2017-07-13) LLVM 4.0.1
- macOS Sierra 10.12.6
- The numbers above seem random.  It usually fails in the 500-1200 range.
- Here's reference information on the `al_run_main` method: [[source]](https://github.com/liballeg/allegro5/blob/8d8a6130d0dc7fd5b7b8750137701e31511808b5/src/allegro.c#L36-L43) [[details]](https://github.com/liballeg/allegro5/blob/8d8a6130d0dc7fd5b7b8750137701e31511808b5/docs/src/refman/misc.txt#L14-L23) [[docs]](http://liballeg.org/a5docs/trunk/getting_started.html#the-main-function).  Allegro requires that you pass your "main" function to theirs as a callback.
- What's really strange is that I have done much more complicated main functions in other examples [[1]](https://github.com/Fryguy/crystal_allegro/blob/master/examples/ex_bitmap.cr) [[2]](https://github.com/Fryguy/crystal_allegro/blob/master/examples/ex_bitmap_flip.cr), and they all work fine.  I was trying to create [this example](https://github.com/Fryguy/crystal_allegro/commit/135db3d870986389dd7a544829583ffff0396e24), and have distilled the problem to the script above.

I'm kind of stuck at this point, and not really sure what else to look for.

### PR DESCRIPTION
**Pull request by asterite**
Fixes #4793
Closes #4793

Please read the doc comments in the commit.

As an example, to implement what is needed in #4793 one can now write:

```crystal
@[Link("allegro")]
lib LibAllegro
  fun al_run_main(argc : LibC::Int, argv : LibC::Char**, user_main : (LibC::Int, LibC::Char** -> LibC::Int)) : LibC::Int
end

# Redefine main function
fun main(argc : Int32, argv : UInt8**) : Int32
  # Pass Crystal's main function as a callback
  LibAllegro.al_run_main(argc, argv, ->Crystal.main)
end
```

This is a breaking change because the macro `redefine_main` is removed.

~One more thing: I expose `main` to Crystal as `__crystal_c_main`. This makes it possible to have a user-defined method named `main`, so kinda fixes what's described in #4761. Next, the compiler should disallow funs and methods with the same names, but I'll send that in a separate PR.~

^ I decided it's better for now to expose it as `main`, because otherwise to redefine it one would need to use the same cryptic name. I'll later change the compiler to never expose `fun main` to Crystal code because it never needs to be invoked.

There's one `TODO` there that is fixed by the first commit in this PR. We can get rid of it after releasing the next version.

I also removed the single usage of `redefine_main`, which maybe was needed before, but removing it I got the SDL samples to work (they didn't work anymore).

### CODE REVIEW

File: src/crystal/main.cr [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+end
+
+module Crystal
+  # Defines the main routine run by normal Crystal programs:
+  #
+  # - Initializes the GC
+  # - Invokes the given *block*
+  # - Handles unhandled exceptions
+  # - Invokes `at_exit` handlers
+  # - Flushes STDOUT and STDERR
```
Comment (Sija) [inline src/crystal/main.cr]: Please, enclose _STDOUT_ and _STDERR_ in backticks `` to link 'em to docs.

File: src/crystal/main.cr [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  # is not setup yet, so nothing that allocates memory
+  # in Crystal (like `new` for classes) can be used.
+  def self.main(argc : Int32, argv : UInt8**)
+    main do
+      _crystal_main(argc, argv)
+    end
+  end
+
+  # Executes the main user code. This normally is executed
+  # after initializing the GC and executing `at_exit` handlers.
```
Comment (Sija) [inline src/crystal/main.cr]: >[...] and **before** executing `at_exit` handlers.

?

File: src/crystal/main.cr [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  #   LibFoo.init_foo_and_invoke_main(argc, argv, ->Crystal.main)
+  # end
+  # ```
+  #
+  # Note that before `Crystal.main` is invoked the GC
+  # is not setup yet, so nothing that allocates memory
+  # in Crystal (like `new` for classes) can be used.
+  def self.main(argc : Int32, argv : UInt8**)
+    main do
+      _crystal_main(argc, argv)
```
Comment (Sija) [inline src/crystal/main.cr]: Shouldn't be here `main_user_code(argc, argv)` instead?

**Comment by ysbaddaden** [comment]
Looks great!

One minor thing: sometimes we don't want a main, for example Android applications are shared objects, loaded from Java (JNI).

**Comment by asterite** [comment]
@ysbaddaden Was that possible before this PR?

File: src/crystal/main.cr [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  def self.main(argc : Int32, argv : UInt8**)
+    main do
+      main_user_code(argc, argv)
+    end
+  end
+
+  # Executes the main user code. This normally is executed
+  # after initializing the GC and before executing `at_exit` handlers.
+  #
+  # You should never invoke this method unless you need to
```
Comment (RX14) [inline src/crystal/main.cr]: Would it be possible to raise early if this is executed twice? Seems like a bit of an edge case but might be worth protecting. This could be another PR though.

File: src/crystal/main.cr [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  #   Crystal.main(argc, argv)
+  #   # some cleanup logic after Crystal main
+  # end
+  # ```
+  #
+  # The `Crystal.main` can also be passed as a callback:
+  #
+  # ```
+  # fun main(argc : Int32, argv : UInt8**) : Int32
+  #   LibFoo.init_foo_and_invoke_main(argc, argv, ->Crystal.main)
```
Comment (Sija) [inline src/crystal/main.cr]: Shouldn't `->Crystal.main` be `->Crystal.main(Int32, UInt8**)`?

**Comment by ysbaddaden** [comment]
@asterite no, it wasn't possible, but since it's touching main declarations, it could have been an opportunity to allow to skip generating a default main function. Anyway, no big deal.

File: src/crystal/main.cr (hunk 5 above)
Comment (asterite) [reply src/crystal/main.cr]: If it's a lib call the compiler figures out the argument types.

Comment (Sija) [reply src/crystal/main.cr]: Didn't know that, cushtie!

**Comment by david50407** [comment]
Can we also provide `Crystal.main(argc : Int32, argv : UInt8**, envp : UInt8**)` in this change?
envp is used in many C programs, maybe we can allow this argument and skip it first.

MSDN page for envp: https://msdn.microsoft.com/en-us/library/k104fy6h.aspx

**Comment by Fryguy** [comment]
@asterite This is awesome, and is exactly what I was trying to accomplish myself, but couldn't figure out.  This is really nice!
